### PR TITLE
Migrate inline_values to path_to_inline_values for package install

### LINKED ISSUE #362: Fully Support for Tanzu Package
### Is your feature request related to a problem? Please describe.

Right now, when installing the Tanzu Package, `inline_values` can only be specified as String or float.
https://registry.terraform.io/providers/vmware/tanzu-mission-control/latest/docs/resources/package_install
https://github.com/vmware/terraform-provider-tanzu-mission-control/blob/001ba2285c941a691629f1c50e1ae5e1756f7fd2/internal/resources/tanzupackageinstall/spec/cluster_spec.go#L71

However, most of the Tanzu Package has structural parameters rather than String or Float.
```
envoy:
 service:
   type: LoadBalancer
   annotations: {}
   externalTrafficPolicy: Cluster
   disableWait: false
 hostPorts:
   enable: true
   http: 80
   https: 443
 hostNetwork: false
 terminationGracePeriodSeconds: 300
 logLevel: info
certificates:
 duration: 8760h
 renewBefore: 360h
```
Therefore, most Tanzu Packages cannot be deployed with the TMC Provider.

### Describe the solution you'd like

Since values is written in YAML, YAML support is desirable.
If this is difficult, Json should be supported.
```
  inline_values = jsonencode({
{
	"certificates": {
		"duration": "8760h",
		"renewBefore": "360h"
	},
	"contour": {
		"configFileContents": {},
		"logLevel": "info",
		"pspNames": "vmware-system-restricted",
		"replicas": 2,
		"useProxyProtocol": false
	},
	"envoy": {
		"hostNetwork": false,
		"hostPorts": {
			"enable": true,
			"http": 80,
			"https": 443
		},
		"logLevel": "info",
		"service": {
			"annotations": {},
			"disableWait": false,
			"externalTrafficPolicy": "Cluster",
			"type": "LoadBalancer"
		},
		"terminationGracePeriodSeconds": 300
	},
	"infrastructure_provider": "vsphere",
	"namespace": "tanzu-system-ingress"
}
  })
```

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ishangupta-ds**
1. **What this PR does / why we need it**:

Fully Support for Tanzu Package by migrating inline_values to path_to_inline_values for package install.

Right now, when installing the Tanzu Package, inline_values can only be specified as String or float. However, most of the Tanzu Package has structural parameters rather than String or Float. Therefore, most Tanzu Packages cannot be deployed with the TMC Provider.

This PR also fixes broken tests for some packaging feature APIs. Namely - `Packages`, `Helm Feature` and `Helm Release`.

2. **Which issue(s) this PR fixes**

Fixes #362 

4. **Additional information**

Since values is written in YAML, YAML support is desirable.

5. **Special notes for your reviewer**

Enhancement / Bug fix PR.

Screenshots:

> Package Install
Screenshot 2024-03-29 at 4 19 26 PM

> Packages
Screenshot 2024-03-28 at 11 10 22 AM

> Helm Feature
Screenshot 2024-03-29 at 9 44 35 AM

> Helm Release
Screenshot 2024-03-29 at 10 42 58 AM

### CODE REVIEW

**Comment by ishangupta-ds** [comment]
[...es/tanzupackageinstall/package_install_resource.go](https://app.codecov.io/gh/vmware/terraform-provider-tanzu-mission-control/pull/386?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=vmware#diff-aW50ZXJuYWwvcmVzb3VyY2VzL3Rhbnp1cGFja2FnZWluc3RhbGwvcGFja2FnZV9pbnN0YWxsX3Jlc291cmNlLmdv) - has acceptance tests

[...esources/tanzupackageinstall/spec/cluster_scope.go](https://app.codecov.io/gh/vmware/terraform-provider-tanzu-mission-control/pull/386?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=vmware#diff-aW50ZXJuYWwvcmVzb3VyY2VzL3Rhbnp1cGFja2FnZWluc3RhbGwvc3BlYy9jbHVzdGVyX3Njb3BlLmdv) - we don't have construct tests across the provider


[...nternal/resources/tanzupackageinstall/spec/spec.go](https://app.codecov.io/gh/vmware/terraform-provider-tanzu-mission-control/pull/386?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=vmware#diff-aW50ZXJuYWwvcmVzb3VyY2VzL3Rhbnp1cGFja2FnZWluc3RhbGwvc3BlYy9zcGVjLmdv) - we don't have construct tests across the provider


[...nal/resources/helmrelease/resource_helm_release.go](https://app.codecov.io/gh/vmware/terraform-provider-tanzu-mission-control/pull/386?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=vmware#diff-aW50ZXJuYWwvcmVzb3VyY2VzL2hlbG1yZWxlYXNlL3Jlc291cmNlX2hlbG1fcmVsZWFzZS5nbw==) - has acceptance tests

[internal/helper/yaml.go](https://app.codecov.io/gh/vmware/terraform-provider-tanzu-mission-control/pull/386?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=vmware#diff-aW50ZXJuYWwvaGVscGVyL3lhbWwuZ28=) - let me know if tests are needed for this file, since it's a simple file reading opration.

File: docs/guides/tanzu-mission-control_policy.md
Comment (vmw-vjn) [inline docs/guides/tanzu-mission-control_policy.md:395]: @ishangupta-ds what is the rationale for introducing padding/space here?

File: docs/data-sources/provisioner.md
Comment (vmw-vjn) [inline docs/data-sources/provisioner.md:1]: @ishangupta-ds what is the rationale for introducing padding/space here?

File: docs/resources/ekscluster.md
Comment (vmw-vjn) [inline docs/resources/ekscluster.md:1]: Is this text change vetted by EKS/LCM team?
Seems like there may be loss of information by these changes which could be important.

File: docs/data-sources/provisioner.md
Comment (ishangupta-ds) [reply docs/data-sources/provisioner.md:1]: It is getting auto generated by the `tfplugindocs` @vmw-vjn

File: docs/resources/ekscluster.md
Comment (ishangupta-ds) [reply docs/resources/ekscluster.md:1]: I see some information has been removed by the docs plugin, I will check with someone from the Terraform team on the usage instructions.

**Review by vmw-vjn — COMMENTED** [review]
Looking at the changes, see text & label changes for many resources beyond DP & Helm. Suggest that appropriate owners of said resources/features be looped for approval - to ensure there isn't any loss of information or intent.

File: docs/resources/ekscluster.md
Comment (ishangupta-ds) [reply docs/resources/ekscluster.md:1]: Removed those changes.

File: docs/data-sources/provisioner.md
Comment (ishangupta-ds) [reply docs/data-sources/provisioner.md:1]: Removed those changes.

**Comment by ishangupta-ds** [comment]
> Looking at the changes, see text & label changes for many resources beyond PackageInstall. Suggest that appropriate owners of said resources/features be looped for approval - to ensure there isn't any loss of information or intent.

Removed all unrelated changes.

File: internal/resources/tanzupackageinstall/package_install_resource.go
Comment (ankitsny) [inline internal/resources/tanzupackageinstall/package_install_resource.go:199]: is this an ignorable error?

Comment (ishangupta-ds) [reply internal/resources/tanzupackageinstall/package_install_resource.go:199]: No its not, so it is being returned in the diagnostics for the same reason.
`diag.FromErr(err)`

Comment (ankitsny) [reply internal/resources/tanzupackageinstall/package_install_resource.go:199]: please consider adding `return` statement, `return diag.FromErr(err)`

File: internal/resources/tanzupackageinstall/spec/cluster_scope.go
Comment (AkbaraliShaikh) [inline internal/resources/tanzupackageinstall/spec/cluster_scope.go:138]: same comment, Why can't we remove this code, and once this change is released this will not be supported.

File: docs/data-sources/package_install.md
Comment (AkbaraliShaikh) [inline docs/data-sources/package_install.md:100]: @ishangupta-ds if it is deprecated, then what is the reason to keep this?
Based on the release version the document will be shown accordingly, right?

File: examples/resources/packageinstall/resource.tf
Comment (AkbaraliShaikh) [inline examples/resources/packageinstall/resource.tf:24]: can you please provide the concrete example with the file as a sample?

File: docs/resources/package_install.md
Comment (AkbaraliShaikh) [inline docs/resources/package_install.md:50]: I don't see the tests are updated with this change? Please add.

File: internal/resources/tanzupackageinstall/spec/cluster_scope.go
Comment (ishangupta-ds) [reply internal/resources/tanzupackageinstall/spec/cluster_scope.go:138]: Users might not be able to update to this new schema in a short span of time so we are following the generic API deprecation flow which we use for TMC as well.

File: docs/data-sources/package_install.md
Comment (ishangupta-ds) [reply docs/data-sources/package_install.md:100]: i will add a deprecated text.

File: examples/resources/packageinstall/resource.tf
Comment (ishangupta-ds) [reply examples/resources/packageinstall/resource.tf:24]: Sure.

File: docs/resources/package_install.md
Comment (ishangupta-ds) [reply docs/resources/package_install.md:50]: I can update current tests to use this new schema.

File: internal/resources/tanzupackageinstall/spec/spec.go
Comment (ramya-bangera) [inline internal/resources/tanzupackageinstall/spec/spec.go:47]: @ishangupta-ds - For deprecation of an optional arg you should follow this procedure https://developer.hashicorp.com/terraform/plugin/framework/deprecations#renaming-an-optional-attribute, have you checked it out?

**Review by vmw-vjn — COMMENTED** [review]
Migrating inline_values to path_to_inline_values seems like a breaking change, to me. Due to which existing TF scripts & states might break once these proposed changes are merged.

Would suggest not removing inline_values completely instead adding as addition field path_to_inline_values to package install; so that existing TF scripts & states in use aren't broken. And future usage can be adapt to use recommended path_to_inline_values.

Please note that final say here is with the feature team. So if FT feel the need to make such breaking changes, please go ahead and confirm intent to handle all customer tickets or feedback that the change brings.

File: internal/resources/tanzupackageinstall/spec/spec.go
Comment (ishangupta-ds) [reply internal/resources/tanzupackageinstall/spec/spec.go:47]: yes

File: internal/resources/tanzupackageinstall/package_install_resource.go
Comment (ishangupta-ds) [reply internal/resources/tanzupackageinstall/package_install_resource.go]: Updated.

File: examples/resources/packageinstall/resource.tf
Comment (ishangupta-ds) [reply examples/resources/packageinstall/resource.tf]: Updated.

File: docs/resources/package_install.md
Comment (ishangupta-ds) [reply docs/resources/package_install.md:50]: Updated.

File: internal/resources/tanzupackageinstall/spec/spec.go
Comment (ishangupta-ds) [reply internal/resources/tanzupackageinstall/spec/spec.go:55]: Updated.

File: docs/data-sources/package_install.md
Comment (ishangupta-ds) [reply docs/data-sources/package_install.md:100]: Added.

**Comment by ishangupta-ds** [comment]
> Migrating inline_values to path_to_inline_values seems like a breaking change, to me. Due to which existing customer Terraform scripts & states might break once these proposed changes are merged.
> 
> Would suggest not removing inline_values completely instead adding as addition field path_to_inline_values to package install; so that existing TF scripts & states in use aren't broken. And future usage can be adapt to use recommended path_to_inline_values.
> 
> Please note that final say here is with the feature team. So if FT feel the need to make such breaking changes, please go ahead and confirm intent to handle all customer tickets or feedback that the change brings.

Sure.

**Review by AkbaraliShaikh — DISMISSED** [review]

**Comment by ishangupta-ds** [comment]
@AkbaraliShaikh / @ramya-bangera  / @vmw-vjn / @ankitsny Please take a look again.

**Comment by ramya-bangera** [comment]
@ishangupta-ds - You plan to merge this to main now? or it should go to the feature branch for tap saas?

**Comment by ishangupta-ds** [comment]
@ramya-bangera these bug fixes / changes are for already released SaaS features via Terraform, not related to the TAP enablement.

**Review by AkbaraliShaikh — APPROVED** [review]

File: docs/data-sources/package_install.md
Comment (ishangupta-ds) [reply docs/data-sources/package_install.md]: Updated.

**Comment by ishangupta-ds** [comment]
Please take a look again. @AkbaraliShaikh / @ramya-bangera

**Review by AkbaraliShaikh — APPROVED** [review]

**Review by ankitsny — APPROVED** [review]

**Review by ramya-bangera — APPROVED** [review]

**Pull request merged**
